### ISSUE TITLE
Incorrect concurrent use of Bunny channel in handler/workers?

### ISSUE THREAD

**Issue by michaeljpetter** [jondot/sneakers#227]
Greetings,

We have started to adopt Sneakers into our system and started by setting up a single worker class with 4 processes with 5 threads each.  After running the worker for a period of time that varies from short to long we observe the workers locking outside of the "work" callback in the worker class.  At this point everything is locked and nothing ever comes back.  This feels like a mutex or shared memory/state issue with multiple threads, so we began digging into the Sneakers code, and we believe we found a significant issue with how Sneakers does threading with regard to Bunny channels.  

Our understanding of the flow is this:
1. A Sneakers Worker class is constructed and it creates a new Sneakers::Queue as well as a thread pool.
2. worker.run is called (by ServerEngine?), which calls queue.subscribe(self) (this is Sneakers::Queue).
3. A shared connection is created if not existing (this seems correct).
4. A **single channel** is created for the worker instance.
5. Subscribe is called on the Bunny::Queue and adds the worker.do_work as a callback to message events.
6. At some time, Bunny receives a message from RabbitMQ and invokes the callback, which invokes worker.do_work (presumably on a Bunny thread).
7. worker.do_work pulls a thread from the thread pool, and runs worker **and handler** code on that thread.

Conclusion: If a worker uses > 1 thread, then worker and handler code will concurrently use the Bunny::Channel, which according to Bunny docs is not supported:

[Concurrency in Bunny](http://rubybunny.info/articles/concurrency.html)
- "Parts of the library (most notably Bunny::Channel) are designed to assume they are not shared between threads."
- "Channels must not be shared between threads."

Per that link, however, it does seem to indicate that simply consuming messages with multiple threads on a single channel **should** work, though not recommended ("It is, however, safe to process deliveries in multiple threads if multi-message acknowledgements are not used."), since Sneakers does not do multi-message acks.

However, this puts serious hidden constraints on **handlers**... which are more likely to need to do other things (e.g., in retry or error scenarios).  Per the link: "publishing from multiple threads on a shared channel may result in frames being sent to RabbitMQ out of order" and "There are other potential conflicts arising from frame interleaving" and I believe that's why we're failing here.  When using the OneShot handler, we see no issues whatsoever.  When a handler (or worker) does anything other than ack (outside of initialize), e.g., declare a queue, publish a message, etc., then we get the frozen/deadlocked worker scenario.

It seems maybe the resolution for our use case could be to do manual thread locking around anything except acks performed in a handler (or even a worker.work method), but it feels like we should not have to explicitly do this in a handler (as a usage pattern).  The Sneakers docs don't mention anything about thread safety in workers or handlers but this seems critically important, so I want to make sure we're not missing anything in the expected usage of Sneakers.  Any thoughts?

**Comment by jondot** [comment]
Hi @michaeljpetter, thanks for taking the time to figure this out.

Indeed the "safe" scenario you mentioned is the one Sneakers is based on, I remember clearing out the concurrency issues with @michaelklishin at the time (not sure if over github, IRC, or newsgroup) and indeed it works exactly as you described. There is contention over a channel but as long as you don't do things such as publish back over different threads you're safe.

So, when you want to reply, you use the embedded publisher, and that's under a mutex already.

Currently I don't know of any deployment bumping into this with the existing codebase (did you make your own handler by a chance?) but I'd love to learn more and help out!

For the docs - there's a lot of learning that's piled up, and I feel Github's Wiki isn't a good platform for that anymore, which is why I'm working on moving documentation to a more static platform (a website, with layout etc.).

**Comment by michaeljpetter** [comment]
Yes, we have created a custom handler, and that is where we are having issues.  If we exclude our custom handler, the deadlocks go away completely.  Our custom handler is a configurable backoff retry handler [`(code is here)`](https://github.com/ionicmobile/sneakers_handlers/blob/new_handlers/lib/sneakers_handlers/configurable_backoff_handler.rb).  (Ignore the 'clear_queues_cache' code; we know this is an issue from the forked code.  Even with that removed, and the queue being declared without expiry, we still have the same concurrency issues.)

> So, when you want to reply, you use the embedded publisher, and that's under a mutex already.

The embedded publisher is available for workers, but I don't see that it's given to the handlers.  It could be brought in manually, I'm sure, but what's interesting is that Sneakers itself does not do that in the MaxRetry handler.  It uses the self-declared @error_exchange, and publishes to it when a message has exceeded the max retry count, ostensibly in a non-threadsafe way.  I would expect the MaxRetry handler to have the same concurrency issues that we are seeing, though it publishes far less often (only on the last failed retry) than our handler (which publishes on every retry), so I would expect a much lower incidence (since publishing concurrently on the channel doesn't _necessarily_ cause deadlock, but it _can_ if the frame order gets jumbled).

Incidentally, even in the embedded publisher, it appears the mutex is used to secure the connection creation and exchange declaration, but **not** the publishing itself, which appears to be subject to the same channel conflicts if the worker is multithreaded.

One key difference in MaxRetry vs our handler is that we are declaring the queues in the handler.  We did test with all queues already existing in RabbitMQ (with all expiries removed) and the same deadlocking occurred, however.  I don't know if this matters, but it is a difference.
